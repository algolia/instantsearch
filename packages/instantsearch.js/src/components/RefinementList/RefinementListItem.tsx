/** @jsx h */

import { h } from 'preact';

import Template from '../Template/Template';

import type { JSX } from 'preact';

export type RefinementListItemProps = {
  facetValueToRefine: string;
  handleClick: (args: {
    facetValueToRefine: string;
    isRefined: boolean;
    originalEvent: MouseEvent;
  }) => void;
  isRefined: boolean;
  subItems?: JSX.Element;
  templateData: Record<string, any>;
  templateKey: string;
  templateProps?: Record<string, any>;
  className: string;
};

function RefinementListItem({
  className,
  handleClick,
  facetValueToRefine,
  isRefined,
  templateProps,
  templateKey,
  templateData,
  subItems,
}: RefinementListItemProps) {
  return (
    <li
      className={className}
      onClick={(originalEvent) => {
        handleClick({
          facetValueToRefine,
          isRefined,
          originalEvent,
        });
      }}
    >
      {/* @MAJOR ensure conformity with InstantSearch.css specs */}
      <Template
        {...templateProps}
        templateKey={templateKey}
        data={templateData}
      />
      {subItems}
    </li>
  );
}

export default RefinementListItem;
