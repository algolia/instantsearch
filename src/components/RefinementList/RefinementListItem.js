import PropTypes from 'prop-types';
import React, { Component } from 'preact-compat';

import Template from '../Template';
import isEqual from 'lodash/isEqual';

class RefinementListItem extends Component {
  componentWillMount() {
    this.handleClick = this.handleClick.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    return !isEqual(this.props, nextProps);
  }

  handleClick(originalEvent) {
    this.props.handleClick({
      facetValueToRefine: this.props.facetValueToRefine,
      isRefined: this.props.isRefined,
      originalEvent,
    });
  }

  render() {
    return (
      <div className={this.props.itemClassName} onClick={this.handleClick}>
        <Template
          data={this.props.templateData}
          templateKey={this.props.templateKey}
          {...this.props.templateProps}
        />
        {this.props.subItems}
      </div>
    );
  }
}

RefinementListItem.propTypes = {
  facetValueToRefine: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  handleClick: PropTypes.func.isRequired,
  isRefined: PropTypes.bool.isRequired,
  itemClassName: PropTypes.string,
  subItems: PropTypes.object,
  templateData: PropTypes.object.isRequired,
  templateKey: PropTypes.string.isRequired,
  templateProps: PropTypes.object.isRequired,
};

export default RefinementListItem;
