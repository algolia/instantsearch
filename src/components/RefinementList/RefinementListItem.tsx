/** @jsx h */

import { h } from 'preact';
import Template from '../Template/Template';

export type RefinementListItemProps = {
  facetValueToRefine: string;
  handleClick: (args: {
    facetValueToRefine: string;
    isRefined: boolean;
    originalEvent: MouseEvent;
  }) => void;
  isRefined: boolean;
  subItems?: h.JSX.Element;
  templateData: Record<string, any>;
  templateKey: string;
  templateProps?: Record<string, any>;
  className: string;
};

function RefinementListItem({
  className,
  handleClick,
  facetValueToRefine,
  isRefined,
  templateProps,
  templateKey,
  templateData,
  subItems,
}: RefinementListItemProps) {
  return (
    <li
      className={className}
      onClick={(originalEvent) => {
        handleClick({
          facetValueToRefine,
          isRefined,
          originalEvent,
        });
      }}
    >
      <Template
        {...templateProps}
        templateKey={templateKey}
        data={templateData}
      />
      {subItems}
    </li>
  );
}

export default RefinementListItem;
