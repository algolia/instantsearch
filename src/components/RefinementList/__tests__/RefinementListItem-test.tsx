/** @jsx h */

import { h } from 'preact';
import { shallow } from 'enzyme';
import { ReactElementLike } from 'prop-types';
import RefinementListItem from '../RefinementListItem';

type Props = {
  facetValue: string;
  facetValueToRefine: string;
  isRefined: boolean;
  handleClick: jest.Mock<any, any>;
  className: string;
  templateData: {
    template: string;
  };
  templateKey: string;
  templateProps: {
    template: string;
  };
  subItems: h.JSX.Element;
};

describe('RefinementListItem', () => {
  const props: Props = {
    facetValue: 'Hello',
    facetValueToRefine: 'wi',
    isRefined: false,
    handleClick: jest.fn(),
    className: 'item class',
    templateData: { template: 'data' },
    templateKey: 'item key',
    templateProps: { template: 'props' },
    subItems: <div />,
  };

  it('renders an item', () => {
    const wrapper = shallow(
      (<RefinementListItem {...props} />) as ReactElementLike
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the right function', () => {
    const out = render(props);
    out.simulate('click');
    expect(props.handleClick).toHaveBeenCalledTimes(1);
  });

  function render(askedProps: Props) {
    return shallow(
      (<RefinementListItem {...askedProps} />) as ReactElementLike
    );
  }
});
