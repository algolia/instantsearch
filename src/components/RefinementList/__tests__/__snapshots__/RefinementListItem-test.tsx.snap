// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefinementListItem renders an item 1`] = `
<li
  className="item class"
  onClick={[Function]}
>
  <Template
    data={
      Object {
        "template": "data",
      }
    }
    rootTagName="div"
    template="props"
    templateKey="item key"
    templates={Object {}}
    templatesConfig={Object {}}
    useCustomCompileOptions={Object {}}
  />
  <div />
</li>
`;
